### ISSUE TITLE
Markdown variables code consumes non-variable parentheticals

### ISSUE THREAD

**Issue by dbolack-ab** [naturalcrit/homebrewery#3371]
### Renderer

v3

### Browser

Chrome

### Operating System

Linux

### What happened?

See attached image.

![chomp](https://github.com/naturalcrit/homebrewery/assets/29744700/0eaa335f-50c8-4f65-82ad-420ff1846ee8)

**Comment by ericscheid** [comment]
Indeed. In Legacy the markup `[this](a plus b)` renders as `<a href="a%20and%20b">this</a>`.

**Comment by ericscheid** [comment]
It's not valid for purposes of the linkref syntax, but even in legacy it works.

` <a href="a%20plus%20b">[this]</a> `

**Comment by 5e-Cleric** [comment]
> It's not valid for purposes of the linkref syntax, but even in legacy it works.
> 
> `<a href="a%20plus%20b">[this]</a>`

i edited my comment, in v3 it output `<p>@[this](this and this)</p>` before variables, no link created

**Comment by ericscheid** [comment]
hmm .. according to https://homebrewery.naturalcrit.com/changelog, it's assigning the contents within the parentheses to the variable and is meant to be immediately outputting. 

Adding `$[this]` to the brew causes `a plus b` to be rendered, showing the assignment has indeed occurred.

<img width="773" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/2605467/ba067f42-cc8a-4b28-b42f-0f8af31c097f">

So .. if the markup is `[this](looks-like-a-possible-url)` it will be processed as a link markup, but if the markup is `[this](spaces are not valid in urls)` then the contents is instead assigned to `$[this]` (and not even output, let alone rendered as a link).

The syntax of `[text](string)` is a collision between the link syntax, and the variable syntax. We screwed up.

**Comment by ericscheid** [comment]
<img width="665" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/2605467/7cff4e36-3ba5-45e1-8c45-d1f1126f946d">

**Comment by 5e-Cleric** [comment]
I actually think that makes a lot of sense to me, i vote we don't change this, there is no usecase for `[this](a and b)`

Other than accessing a folder with spaces, which we should get rid of anyway

**Comment by ericscheid** [comment]
Sometimes someone writes a url like `https://www.domain.com/path/search?q=a plus b`.

Technically, the url should be written as `?q=a%20plus%20b` .. which is what the old link making code of marked would actually do. Most browsers will also do the conversion on the fly, just for convenience.

Alternatively, it might get encoded as `?q=a+plus+b`.

Expecting users to write properly url-encoded strings is just cruel.

**Comment by calculuschild** [comment]
> Expecting users to write properly url-encoded strings is just cruel.

This isn't a Homebrewery thing. This is a Commonmark thing. We aren't imposing some new restriction that wasn't already there before. If people want to have spaces in their URL, there is already a specified way to do this, by enclosing the url in angle brackets: https://spec.commonmark.org/0.31.2/#link-destination

> Indeed. In Legacy the markup `[this](a plus b)` renders as `<a href="a%20and%20b">this</a>`.

This is false. In Legacy, this is still an invalid link, and so is not processed at all, remaining as plain text `[test](a plus b)`.

In short, the *only* change our variables code has made here, is changing the output of invalid inline links. In legacy, the output of an invalid link is the same as the input (it remains unprocessed as `[test](a plus b)`) . With variables, the output of an invalid link is just the link title `[test]`. Also, variable assignment still happens because `[test](a plus b)` is still a valid variable assignment).

**Comment by ericscheid** [comment]
Ah, I suspected there was a change in Marked's linking code that affected V3. That does help explain things more.

Still .. unrelated .. the `[var](value)` syntax is described as "Assigns a variable and outputs it (can be inline)" .. which it clearly is not doing. It is assigning the variable, and only outputting the name of the variable. Only if the value appears to be a valid url does it work (and doing the extra step of transforming the url into a html link) .. which means doing something like `[name](Abraham Lincoln)` just does not behave as it should. 

I believe the intended behaviour is 

1. assign the variable,
2. if valid url then output as a html link
3. if not a valid url then output the value as just text.

Currently, items 1 and 2 work as expected, but item 3 is broken.

**Comment by calculuschild** [comment]
The current behavior is:

1. Split the inline `[label](content)` into two parts. Register an assignment to the processing queue, and drop an emitter`[label]` in its place.
2. Perform the assignment (may involve hoisting other variables, etc.)
3. After all assignments are handled, process any emitters
4. If the emitter has no valid value assigned (was never assigned, etc.) , leave the emitter unprocessed as plain text
5. If the emitter is in link format `[ ]` or image format `![ ]`, check for valid url before emitting
  - If not a valid url, leave the emitter unprocessed as plain text
6. If the emitter is in markdown format `$[ ]`, try to perform math, then emit the result.

All of these steps are executed correctly, which results in equivalent output to this:

```
[label] : And invalid url

[label]
```

This will also perform the assignment, then leave `[label]` untouched because it cannot emit a valid url.

The question, is if we a failed inline `[label](invalid url)` should be displayed as the full original text `[label](invalid url)` (pre-variables behavior), or as `[label]`. 

I could see arguments for both

 - maintain old behavior
vs 
- unify output of invalid variables regardless of inline syntax or not
 
But given that this is just a matter of "what color should the garbage out be when we get garbage in" I don't think it's critical.

**Comment by ericscheid** [comment]
Thanks for the step by step. Yes, I see now that "drop an emitter[label] in its place" explains the behaviour we're seeing.

> But given that this is just a matter of "what color should the garbage out be when we get garbage in" I don't think it's critical to make a the change. We can make the change, but it probably wouldn't be my top priority.

Assigning the value "Abraham Lincoln, Vampire Hunter" to a variable (e.g. `hero-name`) wouldn't be "garbage in". It's assigning a string to a variable.

```
Today we meet the hero of our story: [hero-name](Conan the Barbarian).
He will go on many adventures. Blah blah blah. 
Later we casually name drop here..
[hero-name] awoke, hungry for adventure.
```

The only reason the value is evaluated for url-validity is because the inline-assign-ouput syntax exactly matches the link-text-to-url syntax.

**Comment by calculuschild** [comment]
> `Today we meet the hero of our story: [hero-name](Conan the Barbarian).`

This is invalid syntax. Either it is attempting to generate a link, using and invalid URL, or it is attempting to generate text output, in which case it is missing a `$`: `Today we meet the hero of our story: $[hero-name](Conan the Barbarian).`

Simliarly, `[hero-name] awoke, hungry for adventure.` is also invalid, because you are attempting to emit a link using a variable whose contents are not a valid URL. If you want to emit the string, you are missing the `$`: `$[hero-name] awoke, hungry for adventure.`

**Comment by ericscheid** [comment]
According to the variables documentation in https://homebrewery.naturalcrit.com/changelog, that example is valid. Perhaps the changelog is wrong.

<img width="725" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/2605467/1415b691-ed47-4c90-8a8b-54535d506988">

Changing the example per your suggestions (i.e. prefixing `$` to the inline assignment) also doesn't work. The only way to assign a non-url string to a variable is via the reflink style whole line method.

<img width="1769" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/2605467/fb805fc8-f069-4ca9-832f-f5d2b8af63ac">

> `$[hero-name](Abraham Lincoln, Vampire Hunter)` is valid however. It is assigning a string to a variable and then using $ to select a compatible output format.

Except it doesn't actually work, as seen above.

**Comment by calculuschild** [comment]
Ah, this is because of the `-` in "hero-name".

It is interpreting this as math. Thus, more garbage in. Remove the math symbols and it works.

**Comment by calculuschild** [comment]
> According to the variables documentation in https://homebrewery.naturalcrit.com/changelog, that example is valid. Perhaps the changelog is wrong

You are misreading the changelog. I agree it could be more clear probably, but it seems you missed the line directly above your screenshot, showing that there are 3 cases now, one of which requires a `$`.

The table below shows the common syntax for all 3 cases:

```
[var]:content	
[var](content)
[var]
```

And below, how appending the different prefixes affect the output.

Also see the examples below, which clearly demonstrate the requirement of the $ for all 3 cases.

**Comment by calculuschild** [comment]
> According to the variables documentation in https://homebrewery.naturalcrit.com/changelog, that example is valid. Perhaps the changelog is wrong

You are misreading the changelog. I agree it could be more clear probably, but it seems you missed the line directly above your screenshot, showing that there are 3 cases now, one of which requires a `$`.

The table below shows the common syntax for all 3 cases:

```
[var]:content	
[var](content)
[var]
```

And below, how appending the different prefixes affect the output.

Also see the examples at the bottom, which clearly demonstrate the requirement of the $ for all 3 cases.

**Comment by ericscheid** [comment]
> Ah, this is because of the - in "hero-name". I missed that before.
> It is interpreting this as math.

Doh!

Confirmed. Sheesh.

<img width="1121" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/2605467/879b68c3-482e-4c82-a3fb-85799210fbe4">

> You would see the same issue if you put math symbols in the reflink style whole line method.

Curiously, [same issue] only for non-url values. Hmm .. can't quite wrap my head around all this:

<img width="756" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/2605467/ce865c93-b8d3-4159-bd4b-b29412bb30c9">

Sometimes dashes are allowed, sometimes they break things.

**Comment by dbolack-ab** [comment]
1) This was a new ( and prior to variables addition ) prefix using the same pattern as links. The leading @ should have been relevant.
2) Spaces are absolutely legal in the URL portion of a link declaration. See https://spec.commonmark.org/0.30/#example-481

**Comment by calculuschild** [comment]
> 1) This was a new ( and prior to variables addition ) prefix using the same pattern as links. The leading @ should have been relevant.

Variables are processed before everything else. It sees `@[label](content)` attempts to match a ! or $ prefix, cannot find one, and so moved to the next character and handles it as a standard link. It doesn't consider other extensions until variables are fully processed.

> 2) Spaces are absolutely legal in the URL portion of a link declaration. See https://spec.commonmark.org/0.30/#example-481

Your example has no spaces in the url. The url there is `/url`.

**Comment by dbolack-ab** [comment]
> > 2. Spaces are absolutely legal in the URL portion of a link declaration. See https://spec.commonmark.org/0.30/#example-481
> 
> Your example has no spaces in the url. The url there is `/uri`.

The parentetical. 

This is not worth the frustration. Closing.

**Comment by ericscheid** [comment]
You are both (simultaneously) both right and wrong. 🤣 

There are two relevant examples regarding spaces in the link destination.

<img width="867" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/2605467/e0674e97-c512-4e3a-86d1-bc9d05e08139">

- https://spec.commonmark.org/0.30/#example-487 = doesn't work, renders as brew text
- https://spec.commonmark.org/0.30/#example-488 = how to make it work, renders as html link element

**Comment by calculuschild** [comment]
> - https://spec.commonmark.org/0.30/#example-488 = how to make it work, renders as html link element

LMAO

See my first comment in this issue where I already pointed to the angle bracket exception:
https://github.com/naturalcrit/homebrewery/issues/3371#issuecomment-2012413424

**Comment by calculuschild** [comment]
> Curiously, [same issue] only for non-url values. Hmm .. can't quite wrap my head around all this:

Hm... Yeah this is some odd behavior. For consistency, I think `[any+math-symbols]` should prevent a variable being assigned, so `[ref-05](content)` shouldn't actually assign anything (and thus not express anything either). However I believe the code currently only checks for math when a `$` is present. I will extract this into its own issue.
